kernel/src/clustering.rs

```diff
@@ -0,0 +1,272 @@
+    pub(crate) fn new_with_schema(
+        columns: &[ColumnName],
+        schema: &StructType,
+        column_mapping_mode: ColumnMappingMode,
+    ) -> DeltaResult<Self> {
+        let clustering_columns = columns
+            .iter()
+            .map(|col| {
+                resolve_clustering_column_to_physical_path(col.path(), schema, column_mapping_mode)
+            })
+            .collect::<DeltaResult<Vec<_>>>()?;
+        Ok(Self { clustering_columns })
+    }
+
+    /// Returns the clustering columns as a slice of column paths.
+    #[cfg(test)]
+    pub(crate) fn clustering_columns(&self) -> &[Vec<String>] {
+        &self.clustering_columns
+    }
+
+    /// Converts this clustering metadata to a [`DomainMetadata`] action.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the metadata cannot be serialized to JSON.
+    pub(crate) fn to_domain_metadata(&self) -> DeltaResult<DomainMetadata> {
+        let configuration = serde_json::to_string(self)?;
+        Ok(DomainMetadata::new(
+            CLUSTERING_DOMAIN_NAME.to_string(),
+            configuration,
+        ))
+    }
+
+    /// Creates a clustering metadata domain from a JSON configuration string.
+    ///
+    /// # Arguments
+    ///
+    /// * `json` - The JSON string containing the clustering metadata.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the JSON cannot be parsed.
+    #[cfg(test)]
+    pub(crate) fn from_json(json: &str) -> DeltaResult<Self> {
+        Ok(serde_json::from_str(json)?)
+    }
+}
+
+/// Resolves a single clustering column's logical field path to physical names.
+///
+/// This function is called once per clustering column. It takes a single column's
+/// field path (which may be nested, e.g., `["address", "city"]`) and resolves each
+/// field name to its physical name using the schema's column mapping metadata.
+///
+/// When column mapping is disabled (`None` mode), the logical names are returned as-is.
+///
+/// # Arguments
+///
+/// * `clustering_column_path` - A single clustering column's field path as logical names
+///   (e.g., `["address", "city"]` for a nested column, or `["user_id"]` for a top-level column)
+/// * `schema` - The table schema with column mapping metadata (if enabled)
+/// * `column_mapping_mode` - The column mapping mode (`None`, `Name`, or `Id`)
+///
+/// # Returns
+///
+/// The resolved field path as physical names (if column mapping enabled) or logical names (if disabled).
+///
+/// # Example
+///
+/// Given a schema with column mapping enabled:
+/// ```text
+/// Schema:
+///   - "address" (physicalName: "col-abc123")
+///       - "city" (physicalName: "col-def456")
+///
+/// Input:  clustering_column_path = ["address", "city"]
+/// Output: ["col-abc123", "col-def456"]
+/// ```
+///
+/// The function traverses the schema using logical names to find each field,
+/// then extracts the physical name from the field's metadata.
+fn resolve_clustering_column_to_physical_path(
```

Comment:
+1 that I'm not sure anything here is specific to clustering? seems like a generic column mapping utility? perhaps I'm missing something. I would imagine this would be in its own column mapping utility.

See
- https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java
- https://github.com/delta-io/delta/blob/master/kernel-spark/src/main/java/io/delta/kernel/spark/utils/SchemaUtils.java